pkg/flow/internal/controller/loader.go

```diff
@@ -464,17 +512,57 @@ func (l *Loader) populateComponentNodes(g *dag.Graph, componentBlocks []*ast.Blo
 				})
 				continue
 			}
-
-			// Create a new component
-			c = NewComponentNode(l.globals, registration, block)
+			firstPart := strings.Split(componentName, ".")[0]
+			if l.shouldAddDeclareComponentNode(firstPart, componentName) {
+				g.Add(NewDeclareComponentNode(l.globals, block, l.getModuleInfo))
+			} else {
+				registration, exists := l.componentReg.Get(componentName)
+				if !exists {
+					diags.Add(diag.Diagnostic{
+						Severity: diag.SeverityLevelError,
+						Message:  fmt.Sprintf("Unrecognized component name %q", componentName),
+						StartPos: block.NamePos.Position(),
+						EndPos:   block.NamePos.Add(len(componentName) - 1).Position(),
+					})
+					continue
+				}
+				g.Add(NewNativeComponentNode(l.globals, registration, block))
+			}
 		}
-
-		g.Add(c)
 	}
 
 	return diags
 }
 
+func (l *Loader) shouldAddDeclareComponentNode(firstPart, componentName string) bool {
```

Comment:
This method's name describes more the context where it is called instead of what it does. It's better for methods to describe what they do. This one checks whether a component with given name has been defined via declare - either directly in this config or via imports or in a parent. We should call this something like `isCustomComponent` once we switch to the new naming conventions.